### PR TITLE
[Nickel] Add support for extended patterns

### PR DESCRIPTION
**Pull request by yannham**
# Extended patterns

This PR adds support for the last train of new pattern features in current Nickel: wildcard patterns, or-patterns, array patterns, constant patterns and pattern guards.

## Description

Update to latest Nickel tree-sitter grammar and add queries to format the new constructs.

## Checklist

Checklist before merging:
- [x] CHANGELOG.md updated
- [x] README.md up-to-date

### CODE REVIEW

File: bin/update-wasm-grammars.sh [hunk 1]
```diff
@@ -4,9 +4,10 @@ set -eu
 
 cd web-playground/public/scripts/
 
-WORKDIR="$(mktemp -d)"
+WORKDIR=./tmp/
+mkdir -p $WORKDIR
```
Comment (Xophmeister) [inline bin/update-wasm-grammars.sh:8]: It's worth adding a comment here as to why `mktemp` doesn't work in general.

File: bin/update-wasm-grammars.sh [hunk 2]
```diff
@@ -4,9 +4,10 @@ set -eu
...
 readonly WORKDIR
-trap 'echo -e "${BLUE}Cleanup...${NC}"; rm -rf "${WORKDIR}"' EXIT
+trap 'echo -e "${BLUE}Cleanup...${NC}"; rm -rf "${WORKDIR}"; rm a.out.{js,wasm}' EXIT
```
Comment (Xophmeister) [inline bin/update-wasm-grammars.sh:10]:
```suggestion
trap 'echo -e "${BLUE}Cleanup...${NC}"; rm -rf "${WORKDIR}" a.out.{js,wasm}' EXIT
```

**Review by Xophmeister — CHANGES_REQUESTED** [review]
The changes look good, but I would suggest:

@yannham Can some example code that hit your new formatting rules be added to `topiary-cli/tests/samples/{input,expected}/nickel.ncl` so they are tested?

@ErinvanderVeen I would cherry-pick your changes to `bin/update-wasm-grammars.sh` into a separate PR -- along with the updates to the _non_-Nickel WASM grammars -- so this PR can be more focused.

**Comment by ErinvanderVeen** [comment]
> @yannham Can some example code that hit your new formatting rules be added to `topiary-cli/tests/samples/{input,expected}/nickel.ncl` so they are tested?

The new tests were blocked because there was an issue in the tree-sitter grammar. Yann has resolved this issue, but it still needs to merged.

**Comment by yannham** [comment]
If I understood correctly, @ErinvanderVeen is willing to take over and to take this PR to the finish line.

**Comment by yannham** [comment]
In fact I'm going to take over again, as this is blocking for the next Nickel release :slightly_smiling_face:

**Comment by Xophmeister** [comment]
Looks good... Let me just cherry-pick @ErinvanderVeen's commit into a separate branch before it gets lost. Also, I'll regenerate the Nickel WASM for this PR. Give me a sec, then I'll approve :+1:

**Review by Xophmeister — APPROVED** [review]

**Pull request merged**